Show whole post on short posts

RyanTG opened the pull request:
## Description

Adds logic to content-list html to allow users to show the full post on the content-list if the post has fewer than 500 characters.

## Motivation and Context

Currently, if you have a short post - say a post that's just an image, or one with only a few words - then you still get a "summary" of it in the content list, with a "read more" button. This gives the user the option to show the full post if the post has fewer than 500 characters.

## Checklist:

- [ ] I have updated the [documentation](https://github.com/pacollins/hugo-future-imperfect-slim/wiki), as applicable.
- [x] I have updated the `theme.toml`, as applicable.

I updated the example site config with the new param. And left a comment describing it. If preferable, I can remove the comment and/or add the info to the wiki.

This ([and my other PR](https://github.com/pacollins/hugo-future-imperfect-slim/pull/174)) should not be showing those extra commits... I hate it when github does this. I will look into resolving it. But it shouldn't affect merging.

pacollins commented on exampleSite/config.toml:
```diff
@@ -27,6 +27,7 @@ disableLanguages        = []
   readingTime           = true
   imageStretch          = ""
   socialShare           = ["twitter", "facebook", "reddit", "linkedin", "pinterest", "email"]
+  readMoreHide          = true # hide "read more" if post is under 500 characters, and show whole post
```
1. Default values should not change the theme for current users.
2. I think, based on the other parameters, `hideReadMore` would match better. 
3. Also, please add the description to the wiki, or leave it in the PR post, and I will add it to the wiki when I merge.

pacollins commented on layouts/_default/content-list.html:
```diff
@@ -2,11 +2,19 @@
     {{ .Render "featured" }}
-    {{ $.Scratch.Set "summary" ((delimit (findRE "<p.*?>(.|\n)*?</p>" .Content 1) "") | truncate (default 500 .Site.Params.summary_length) (default "&hellip;" .Site.Params.text.truncated ) | replaceRE "&amp;" "&" | safeHTML) }}
-    {{ $.Scratch.Get "summary" }}
+    {{ $.Scratch.Set "summary" ((delimit (findRE "<p.*?>(.|\n)*?</p>" .Content 2) "") | truncate (default 500 .Site.Params.summary_length) (default "&hellip;" .Site.Params.text.truncated ) | replaceRE "&amp;" "&" | safeHTML) }}
```
`summary_length` was looks like it was either [wrong originally or deprecated](https://gohugo.io/getting-started/configuration). Please verify and change to `summaryLength` as needed.

pacollins commented on layouts/_default/content-list.html:
```diff
@@ -2,11 +2,19 @@
-    {{ $.Scratch.Set "summary" ((delimit (findRE "<p.*?>(.|\n)*?</p>" .Content 1) "") | truncate (default 500 .Site.Params.summary_length) (default "&hellip;" .Site.Params.text.truncated ) | replaceRE "&amp;" "&" | safeHTML) }}
-    {{ $.Scratch.Get "summary" }}
+    {{ $.Scratch.Set "summary" ((delimit (findRE "<p.*?>(.|\n)*?</p>" .Content 2) "") | truncate (default 500 .Site.Params.summary_length) (default "&hellip;" .Site.Params.text.truncated ) | replaceRE "&amp;" "&" | safeHTML) }}
+    {{ if and (eq .Site.Params.readMoreHide true) (ge (len .Content) 500) }}
```
This, and following lines, should reference `summaryLength` from above with a default value of 500.

RyanTG commented:
I have addressed the comments. Thanks! 

I'm not sure how to edit the wiki. Pull it and then make a separate PR? It's very unclear...

Here is the description:

hideReadMore
    Hides "read more" for a post on the list page if post is under 500 characters, and instead shows the whole post.

pacollins commented:
Now that I've had you fiddle with this and I've spent time digging in the Hugo Docs, I have a question on the intent.

1. If a post has 500 characters or less, it would display the entire post.
2. If a post has 501 characters or more, it would display a 70 word summary.

I think what we really want to do is make people aware of the [`summaryLength`](https://gohugo.io/content-management/summaries/#automatic-summary-splitting) (which is placed under the main site config, check for [`.Truncated`](https://gohugo.io/variables/page/), and then adjust `content-list.html` to:

```
<article class="post">
  {{ .Render "header" }}
  <div class="content">
    {{ .Render "featured" }}
    {{ .Summary | safeHTML }}
    <!-- We could add an optional elipse here, or something, but I don't know that it is needed -->
  </div>
  <footer>
    {{ if .Truncated }}
      <a href="{{ .RelPermalink }}" class="button big">{{ i18n "read_more" }}</a>
    {{ end }}
    {{ .Render "stats" }}
  </footer>
</article>
```

This will actually significantly clean the code and reduce specificity.

However, if your intent is to have the summary short unless a post is so long, then we should continue that discussion.

RyanTG commented:
In my latest commit - https://github.com/pacollins/hugo-future-imperfect-slim/pull/177/commits/b3a77f7d41ad89cca9624e32af893722fe089f4f - I added summaryLength to config, since we didn't have it, and I thought 70 "words" was far too few. And I put "words" in quotes, because while I saw that in the docs, when I used summaryLength with the default value, it seemed to be truncating the post at 70 _characters_, not words.

That said, I agree with you here. This is much cleaner, and I believe it accomplishes my goals. I will work with this idea, and update this PR.

pacollins commented:
When I was testing early, it was doing words. I do not know that we need to explicitly include it in the `config.toml` since it is a Hugo parameter, but it might be worthwhile to add it to the wiki. I will do that as I update for #176.

If you know how, try to rebase your PR since I merged that other and it created conflicts.

RyanTG commented:
I'm going to spend more time with this, but my thoughts so far.

# 1) The summary output is ugly

`.Summary` strips html (unless you use `<!--more-->`), which I think makes for ugly, more difficult to read summaries. 

For example, I have a post that begins with a header tag, a short paragraph, and then another header tag.

````
## Goal

The goal here is to filter OpenStreetMap data extracts to only the features you need to use. The result is a smaller, more manageable dataset.

## Background

For my [map wall art][1] I'm making maps of many of the major cities in the United States (and will do other areas once I'm done with that). To be efficient, I'm using continent-wide datasets from OpenStreetMap (so I can just zoom to an area and export PDFs). My maps are fairly minimal, and the datasets contain a lot of data that I don't need. This post is a simple tutorial about how to do an SQL query of the data _as_ you convert it to a database table in order to reduce the size of the output. This isn't a secret or anything, but I only found a couple examples of this process, so I thought I'd add my own.
````
Our current summary removes the headers, and retains the paragraphs:
````
The goal here is to filter OpenStreetMap data extracts to only the features you need to use. The result is a smaller, more manageable dataset.

For my map wall art I’m making maps of many of the major cities in the United States (and will do other areas once I’m done with that). To be efficient, I’m using continent-wide datasets from OpenStreetMap (so I can just zoom to an area and export PDFs). My maps are fairly minimal, and the datasets contain a lot of data that I don’t…
````

But `.Summary` turns it into this:

````
Goal The goal here is to filter OpenStreetMap data extracts to only the features you need to use. The result is a smaller, more manageable dataset. Background For my map wall art I’m making maps of many of the major cities in the United States (and will do other areas once I’m done with that). To be efficient, I’m using continent-wide datasets from OpenStreetMap (so I can just zoom to an area and export PDFs).
````
Which, I think is unacceptable.

# 2) If you have a post with an image, but no text, the image won't show.

This is kind of an outlier example. But I have posts that are simply comics I've made, with no text. And the image is _not_ a featured image from the front matter, but is instead in the post body. The `.Summary` completely skips the image, and the post in the content list shows as blank, with no content.

Again, this can be solved by using `<!--more-->

# Thoughts?

RyanTG commented:
Actually, this works pretty dang well `{{ .Content | safeHTML | truncate 500 }}`

pacollins commented:
So `{{ .Summary | safeHTML }}` does not work correctly?

On Tue, Aug 4, 2020 at 12:54 AM Ryan TG <notifications@github.com> wrote:

> Actually, this works pretty dang well {{ .Content | safeHTML | truncate
> 500 }}
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pacollins/hugo-future-imperfect-slim/pull/177#issuecomment-668378569>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACWVJSMBSUAWZHFR7P25ZNLR66H6ZANCNFSM4PJICNGQ>
> .
>

RyanTG commented:
I wish it did, since then we could use summaryLength. But as it is, no.

RyanTG commented:
@pacollins Just to be clear, because I think you’re reading and responding via email and might be missing messages and edits to messages: I want to make sure you saw the comment where I gave a couple detailed examples of why .Summary isn’t great https://github.com/pacollins/hugo-future-imperfect-slim/pull/177#issuecomment-668376392

Your last comment implied you hasn’t seen it.

RyanTG commented:
Ok, I think this is less disruptive to our current theme users (who may not being using `<!--more-->`.

pacollins commented:
At least locally, `<!--more-->` no longer functions as intended.

Additionally, by showing HTML and truncating by character, we have issues like this:
![image](https://user-images.githubusercontent.com/11359433/89702582-b2c23e80-d910-11ea-9181-f1e78c2ce903.png)

RyanTG commented:
Ah. Yeah, that's no good. And I'm beginning to think this idea might not be a good one for the theme, because even when you're "showing the whole post" on content-list, you're still missing elements like the social buttons (I happen to have those social buttons disabled on my personal site). And it won't be obvious to users that they need to click the title in order to see those buttons and to leave/view comments. 

I'm fine with closing this if you think it's a lost cause. 

On my personal site I'm actually going to keep plugging away at it. So far I'm at:

````
<article class="post">
  {{ .Render "header" }}
  <div class="content">
    {{ .Render "featured" }}
    {{ if le ( .WordCount ) .Site.Params.summaryLength }}
      {{ .Content }}
    {{ else }}
      {{ $.Scratch.Set "summary" ((delimit (findRE "<p.*?>(.|\n)*?</p>" .Content 1) "") | truncate (default 500 .Site.Params.summaryLength) (default "&hellip;" .Site.Params.text.truncated ) | replaceRE "&amp;" "&" | safeHTML) }}
      {{ $.Scratch.Get "summary" }}
    {{ end }}
  </div>
  <footer>
    {{ if gt ( .WordCount ) .Site.Params.summaryLength }}
      <a href="{{ .RelPermalink }}" class="button big">{{ i18n "read_more" }}</a>
    {{ end }}
    {{ .Render "stats" }}
  </footer>
</article>

````

This preserves the site's current summary formatting (either cut after first paragraph, or use summary), and allows the user to set a summaryLength to customize things. 

Logic:
1. If the content length (using .WordCount, since that's more a "pure" count of words, vs .Content which also counts html) is equal to the summary length, then that means it's a short post, and we should show the whole post.

2. In the footer, we can't use `if .Truncated` because one of the summary options is to delimit after the first paragraph (which is not a "truncation"). Additionally, I will add that I do not like using .summaryLength in the truncate, because _I swear_ it uses characters and not words in that instance, while elsewhere it uses words. I don't see how that is possible, but that is what I am seeing.

pacollins commented:
Feel free to continue to pursue and/or use this for your personal version of theme (just make sure to relocate the file so you can get updates correctly).

I am going to mark this as "In Progress" and feel free to close at any time if you determine it is not feasible to solve the issues addressed above..
